src/jrd/tra.cpp

```diff
@@ -3651,11 +3652,45 @@ void jrd_tra::checkBlob(thread_db* tdbb, const bid* blob_id)
 			if (blb_relation->rel_security_name.isEmpty())
 				MET_scan_relation(tdbb, blb_relation);
 			SecurityClass* s_class = SCL_get_class(tdbb, blb_relation->rel_security_name.c_str());
-			if (s_class && !s_class->scl_blb_access)
+			if (s_class)
 			{
-				SCL_check_access(tdbb, s_class, 0, 0, NULL, SCL_select, SCL_object_table, false, 
-					blb_relation->rel_name);
-				s_class->scl_blb_access = true;
+				if (s_class->scl_blb_access == SecurityClass::BA_UNKNOWN)
+				{
+					// Relation has not been checked for access rights
+					try
+					{
+						SCL_check_access(tdbb, s_class, 0, 0, NULL, SCL_select, SCL_object_table, false,
+							blb_relation->rel_name);
+						s_class->scl_blb_access = SecurityClass::BA_SUCCESS;
+					}
+					catch (const Exception&)
+					{
+						if (tdbb->tdbb_status_vector->getErrors()[1] != isc_no_priv)
+							throw;
+
+						// We don't have access to this relation
+						s_class->scl_blb_access = SecurityClass::BA_FAILURE;
+
+						if (punt)
+							throw;
+
+						// but someone else has (SP, view)
+						// store Blob ID as allowed in this transaction
+						tra_fetched_blobs.add(*blob_id);
+						tdbb->tdbb_status_vector->clearException();
+					}
+				}
```

Comment:
Why not use switch() here instead if / else if ?

Reply:
Did not notice this when reworked the logic.